### PR TITLE
macos arm64 support

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by isuruf** [comment]
ping @matthew-brett, @mattip, @rgommers

File: gfortran_utils.sh [hunk 1]
```diff
@@ -97,6 +97,21 @@ if [ "$(uname)" == "Darwin" ]; then
         hdiutil attach -mountpoint /Volumes/gfortran $GFORTRAN_DMG
         sudo installer -pkg /Volumes/gfortran/gfortran.pkg -target /
         check_gfortran
+        if [[ "${PLAT:-}" == "universal2" || "${PLAT:-}" == "arm64" ]]; then
```
Comment (matthew-brett) [inline gfortran_utils.sh:122]: Just checking - do we also need the `gfortran.pkg` above?

Do we need to add an archive for this compiler as well?

Comment (isuruf) [reply gfortran_utils.sh:122]: > Just checking - do we also need the gfortran.pkg above?

For some rare cases where for cross compiling you need a native compiler to generate some executables that are used during the build.

> Do we need to add an archive for this compiler as well?

I don't understand. Can you clarify?

File: gfortran_utils.sh [hunk 2]
```diff
@@ -97,6 +97,21 @@ if [ "$(uname)" == "Darwin" ]; then
...
+            curl -L -O https://github.com/isuruf/gcc/releases/download/gcc-10-arm-20210112/gfortran-darwin-arm64.tar.gz
```
Comment (matthew-brett) [inline gfortran_utils.sh]: Maybe encapsulate this as its own function - like `install_arm64_gfortran`?

File: gfortran_utils.sh [hunk 3]
```diff
@@ -97,6 +97,21 @@ if [ "$(uname)" == "Darwin" ]; then
...
+            sudo mkdir -p /opt/
+            sudo cp gfortran-darwin-arm64.tar.gz /opt/gfortran-darwin-arm64.tar.gz
+            pushd /opt
+                sudo tar -xvf gfortran-darwin-arm64.tar.gz
+                sudo rm gfortran-darwin-arm64.tar.gz
+            popd
+            export FC_ARM64="/opt/gfortran-darwin-arm64/bin/arm64-apple-darwin20.0.0-gfortran"
```
Comment (matthew-brett) [inline gfortran_utils.sh]: Any way of getting this path more automatically?  Or generalizing so we don't have to change it with other releases?

File: gfortran_utils.sh (hunk 1 above)
Comment (matthew-brett) [reply gfortran_utils.sh:122]: Or are the releases stored somewhere in version control - within Github?

**Review by matthew-brett — COMMENTED** [review]
A few immediate thoughts.

File: gfortran_utils.sh (hunk 1 above)
Comment (matthew-brett) [reply gfortran_utils.sh:122]: As you can see - I stored the binary blob for the installer in version control here to be sure we had it, and we knew exactly what it was.  The question is how to do the same for your gcc toolchain.  Can we refer to your repository commit or similar?  Are the binary artefacts guaranteed to be available for the medium term, for reproducible builds?

Comment (isuruf) [reply gfortran_utils.sh:122]: Checked into this repo

File: gfortran_utils.sh (hunk 3 above)
Comment (isuruf) [reply gfortran_utils.sh]: Since the archive is checked into the repo, this won't change.

File: gfortran_utils.sh (hunk 2 above)
Comment (isuruf) [reply gfortran_utils.sh]: Done

File: gfortran_utils.sh [hunk 4]
```diff
@@ -93,10 +93,28 @@ if [ "$(uname)" == "Darwin" ]; then
...
+        popd
+    }
+
     function install_gfortran {
         hdiutil attach -mountpoint /Volumes/gfortran $GFORTRAN_DMG
         sudo installer -pkg /Volumes/gfortran/gfortran.pkg -target /
         check_gfortran
+        if [[ "${PLAT:-}" == "universal2" || "${PLAT:-}" == "arm64" ]]; then
+            install_arm64_cross_gfortran
+            export FC_ARM64="/opt/gfortran-darwin-arm64/bin/arm64-apple-darwin20.0.0-gfortran"
```
Comment (matthew-brett) [inline gfortran_utils.sh]: Can't these exports go in the function?

File: gfortran_utils.sh (hunk 3 above)
Comment (matthew-brett) [reply gfortran_utils.sh]: Well - sorry - I'm not saying it's necessarily a good idea to check it into the repo - if there is some other way to guarantee that we can get exactly that binary from something similar to version control elsewhere, that would be just as good.

**Review by matthew-brett — COMMENTED** [review]
A few more.

File: gfortran_utils.sh (hunk 3 above)
Comment (isuruf) [reply gfortran_utils.sh]: Reverted back to the URL and added a shasum check.

File: gfortran_utils.sh [hunk 5]
```diff
@@ -93,10 +93,34 @@ if [ "$(uname)" == "Darwin" ]; then
     GFORTRAN_DMG="${GF_UTIL_DIR}/archives/gfortran-4.9.0-Mavericks.dmg"
     GFORTRAN_SHA="$(shasum $GFORTRAN_DMG)"
 
+    function install_arm64_cross_gfortran {
+        curl -L -O https://github.com/isuruf/gcc/releases/download/gcc-10-arm-20210112/gfortran-darwin-arm64.tar.gz
```
Comment (matthew-brett) [inline gfortran_utils.sh]: Right - but just to check - do you think we can depend on this curl command working for the next few years?

Comment (isuruf) [reply gfortran_utils.sh]: Sure.

Comment (isuruf) [reply gfortran_utils.sh]: We can also make a release in `MacPython/gfortran-install` and attach it there.

Comment (matthew-brett) [reply gfortran_utils.sh]: Yes, maybe that would be the best way.

**Comment by matthew-brett** [comment]
@mattip - any thoughts?

**Comment by mattip** [comment]
Does it make any difference to CI performance if the file is part of this repo or if it is downloaded from a github release URL? In my mind I would hop somehow subrepos can be cached, but the URL must be fetched each run, but I don't know if this intuition is true or not.

**Comment by isuruf** [comment]
No, downloading took 2 seconds on github actions for me.

**Comment by mattip** [comment]
Cool. Then LGTM.

**Comment by matthew-brett** [comment]
Thanks @isuruf

**Pull request merged**